### PR TITLE
Add some debug log to OpamCudf.extract_explanations to help debug #4373

### PR DESCRIPTION
**Pull request by kit-ty-kate**
The following dockerfile (derived from https://github.com/ocaml/opam/issues/4373#issuecomment-992725743):
```
FROM ocaml/opam:debian-11-ocaml-4.12@sha256:40813dc1391636b4a428afbf0e678d2c81b80d9eade8a5214d8ee271fae155f3
RUN git clone https://github.com/kit-ty-kate/opam.git -b extract_explanations-debug
WORKDIR opam
RUN opam exec -- ./configure --with-vendored-deps
RUN opam exec -- make
RUN sudo ln -f _build/default/src/client/opamMain.exe /usr/bin/opam
RUN opam init --reinit -ni
RUN opam remote set-url default git+https://github.com/ocaml/opam-repository.git#a5d7cdc0c91452b0aef4fa71c331ee5237f6dddd
RUN opam pin add --debug --debug-level=3 --with-version 0.6.1 git+https://github.com/bikallem/resto.git#e725f56dd2d5ca0148272d754a0f29ffe76f94e8
```
now reports:
```
00:10.228  CLIENT                          conflict!
[ERROR] Package conflict!
00:10.228  CUDF                            Conflict reporting
00:10.228  CUDF                            Reasons: [
  Dependency (dose-dummy-request.???, [resto-cohttp-server], [resto-cohttp-server.0.6.1]);
  Dependency (resto-cohttp-server.0.6.1, [cohttp-lwt-unix>=4.0.0], [cohttp-lwt-unix.4.0.0]);
  Dependency (resto-cohttp-server.0.6.1, [conduit-lwt-unix>=5.0.0], [conduit-lwt-unix.5.0.0]);
  Conflict (conduit-lwt-unix.1.0.3, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.1.4.0, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.4.0.1, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.1.5.0, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.1.2.0, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.2.0.1, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.1.0.2, conduit-lwt-unix.5.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.2.3.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.4.0.2, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.2.2.2, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.4.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.2.0.2, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.2.0.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.1.1.0, conduit-lwt-unix);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.1.3.0, conduit-lwt-unix);
  Dependency (cohttp-lwt-unix.4.0.0, [conduit-lwt-unix<5.0.0], [conduit-lwt-unix.1.2.0, conduit-lwt-unix.1.0.2, conduit-lwt-unix.2.0.1, conduit-lwt-unix.4.0.0, conduit-lwt-unix.4.0.2, conduit-lwt-unix.2.2.2, conduit-lwt-unix.2.3.0, conduit-lwt-unix.1.4.0, conduit-lwt-unix.1.0.3, conduit-lwt-unix.1.5.0, conduit-lwt-unix.1.3.0, conduit-lwt-unix.2.0.2, conduit-lwt-unix.4.0.1, conduit-lwt-unix.1.1.0, conduit-lwt-unix.2.1.0, conduit-lwt-unix.2.0.0]);
  Conflict (conduit-lwt-unix.5.0.0, conduit-lwt-unix.2.1.0, conduit-lwt-unix);
]
[ERROR] Internal error while computing conflict explanations: sorry, please report.
[NOTE] Pinning command successful, but your installed packages may be out of sync.
```

### CODE REVIEW

**Comment by AltGr** [comment]
Indeed that repro case is really small, which is very good!

**Review by rjbou — APPROVED** [review]

File: src/solver/opamCudf.ml [hunk 1]
```diff
@@ -968,7 +1045,11 @@ let extract_explanations packages cudfnv2opam reasons : explanation list =
         | `Missing (_, sdeps', fdeps') -> sdeps = sdeps' && fdeps = fdeps'
         | _ -> false)
   in
+  log ~level:3 "Explanations: %a" Pp_explanation.pp_explanationlist explanations;
   match explanations with
+  | [] ->
+    OpamConsole.error_and_exit `Internal_error
+      "Internal error while computing conflict explanations: sorry, please report."
```
Comment (dra27) [inline src/solver/opamCudf.ml:1052]: This wants to be more explicit, especially if something were to happen where a lot of users suddenly see it. The best thing is probably to direct users to an/the issue page, which can then be closed as and when we manage to eliminate the problem!

Comment (kit-ty-kate) [reply src/solver/opamCudf.ml:1052]: (adding comments from https://github.com/ocaml/opam/pull/4982#discussion_r779385255)

Comment (kit-ty-kate) [reply src/solver/opamCudf.ml]: Is it better now?

File: src/solver/opamCudf.ml [hunk 2]
```diff
@@ -968,7 +1045,13 @@ let extract_explanations packages cudfnv2opam reasons : explanation list =
         | `Missing (_, sdeps', fdeps') -> sdeps = sdeps' && fdeps = fdeps'
         | _ -> false)
   in
+  log ~level:3 "Explanations: %a" Pp_explanation.pp_explanationlist explanations;
   match explanations with
+  | [] ->
+    OpamConsole.error_and_exit `Internal_error
+      "Internal error while computing conflict explanations:\n\
+       sorry about that. Please report how you got here in \
+       https://github.com/ocaml/opam/issues/4373 if possible."
```
Comment (rjbou) [inline src/solver/opamCudf.ml:1054]: I don't agree having such url in an error message. Just redirecting them to open an issue is sufficient, or point to the issue tracker if you want to set an url.

**Review by rjbou — CHANGES_REQUESTED** [review]

File: src/solver/opamCudf.ml (hunk 1 above)
Comment (rjbou) [reply src/solver/opamCudf.ml]: x-ref https://github.com/ocaml/opam/pull/4981#discussion_r789162066

File: src/solver/opamCudf.ml (hunk 2 above)
Comment (kit-ty-kate) [reply src/solver/opamCudf.ml]: how is it now?

File: src/solver/opamCudf.ml [hunk 3]
```diff
@@ -968,7 +1045,13 @@ let extract_explanations packages cudfnv2opam reasons : explanation list =
         | `Missing (_, sdeps', fdeps') -> sdeps = sdeps' && fdeps = fdeps'
         | _ -> false)
   in
+  log ~level:3 "Explanations: %a" Pp_explanation.pp_explanationlist explanations;
   match explanations with
+  | [] ->
+    OpamConsole.error_and_exit `Internal_error
+      "Internal error while computing conflict explanations:\n\
+       sorry about that. Please report how you got here in \
+       the bugtracker if possible."
```
Comment (rjbou) [inline src/solver/opamCudf.ml:1054]:
```suggestion
    OpamConsole.error_and_exit `Internal_error
      "Internal error while computing conflict explanations:\n\
       sorry about that. Please report how you got here in \
       https://github.com/ocaml/opam/discussions/5130 if possible"
```

**Review by rjbou — APPROVED** [review]

**Pull request merged**